File: 백준/Silver/14940. 쉬운 최단거리/쉬운 최단거리.c
#include <stdio.h>
#define SIZE 1000000

typedef struct{
    int x;
    int y;
}P;
int graph[1001][1001];
int visited[1001][1001] = {0};
int n, m;
P queue[SIZE];
int front = 0, rear = 0;
int src_x, src_y;
int none = 0;

void push(P data){
    rear = (rear + 1) % SIZE;
    queue[rear] = data;
}
P pop(){
    front = (front + 1) % SIZE;
    return queue[front];
}

int dx[4] = {-1, 1, 0, 0};
int dy[4] = {0, 0, -1, 1};

void BFS(){
    P src;
    src.x = src_x;
    src.y = src_y;
    push(src);
    graph[src_x][src_y] = 0;
    visited[src_x][src_y] = 1;

    while(front != rear){
        P cur = pop();
        for(int i = 0; i < 4; i++){
            int nx = cur.x + dx[i];
            int ny = cur.y + dy[i];
            if(nx < 0 || ny < 0 || nx > n - 1 || ny > m - 1) continue;
            if(!visited[nx][ny] && graph[nx][ny] == 1){
                P new;
                new.x = nx;
                new.y = ny;
                push(new);
                graph[nx][ny] = graph[cur.x][cur.y] + 1;
                visited[nx][ny] = 1;
            }
        }
    }
    return;
}

int main(void){
    int i, j;
    scanf("%d %d", &n, &m);
    for(i = 0; i < n; i++)
        for(j = 0; j < m; j++){
            scanf("%d", &graph[i][j]);
            if(graph[i][j] == 2){
                src_x = i;
                src_y = j;
            }
        }
    BFS();
    for(i = 0; i < n; i++){
        for(j = 0; j < m; j++){
            if(!visited[i][j] && graph[i][j])
                printf("-1 ");
            else
                printf("%d ", graph[i][j]);
        }
        printf("\n");
    }
}

File: 백준/Silver/14940. 쉬운 최단거리/쉬운 최단거리.java
import java.io.*;
import java.util.*;

public class Main {

    boolean[][] map = new boolean[1001][1001];
    int[][] visited = new int[1001][1001];
    int[] dx = {0, 0, -1, 1};
    int[] dy = {-1, 1, 0, 0};
    int N, M;

    public class Point{
        int x;
        int y;
        int dis;
        public Point(int x, int y, int dis){
            this.x = x;
            this.y = y;
            this.dis = dis;
        }
    }

    public void bfs(int x, int y){
        for(int i = 1; i <= N; i++) Arrays.fill(visited[i], -1);
        ArrayDeque<Point> queue = new ArrayDeque<>();
        queue.addLast(new Point(x, y, 0));

        while(!queue.isEmpty()){
            Point now = queue.removeFirst();
            if(visited[now.x][now.y] != -1) continue;
            visited[now.x][now.y] = now.dis;

            for(int i = 0; i < 4; i++){
                int nx = now.x + dx[i];
                int ny = now.y + dy[i];
                if(nx < 1 || ny < 1 || nx > N || ny > M || visited[nx][ny] != -1 || !map[nx][ny]) continue;
                queue.addLast(new Point(nx, ny, now.dis + 1));
            }
        }
    }

    public void solution() throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        StringBuilder sb = new StringBuilder();

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        int srcX = 0, srcY = 0;
        for(int i = 1; i <= N; i++){
            st = new StringTokenizer(br.readLine());
            for(int j = 1; j <= M; j++){
                int num = Integer.parseInt(st.nextToken());
                if(num == 2){
                    srcX = i;
                    srcY = j;
                }else if(num == 1) map[i][j] = true;
            }
        }
        bfs(srcX, srcY);
        for(int i = 1; i <= N; i++){
            for(int j = 1; j <= M; j++){
                if(visited[i][j] == -1 && !map[i][j]) sb.append("0 ");
                else sb.append(visited[i][j]).append(' ');
            }
            sb.append('\n');
        }
        System.out.print(sb);
    }
    public static void main(String[] args) throws Exception {
        new Main().solution();
    }
}